src/main/java/io/cdap/plugin/gcp/common/GCPUtils.java

```diff
@@ -49,11 +64,55 @@ public static ServiceAccountCredentials loadServiceAccountCredentials(String pat
     }
   }
 
+  public static ServiceAccountCredentials loadServiceAccountCredentials(String content,
+                                                                        boolean isServiceAccountFilePath)
+    throws IOException {
+    if (isServiceAccountFilePath) {
+      return loadServiceAccountCredentials(content);
+    }
+    InputStream jsonInputStream = new ByteArrayInputStream(content.getBytes());
+    return ServiceAccountCredentials.fromStream(jsonInputStream);
+  }
+
+  public static String extractPrivateKey(ServiceAccountCredentials credentials) {
+    return String.format(PRIVATE_KEY_WRAP,
+                         Base64.getEncoder().encodeToString(credentials.getPrivateKey().getEncoded()));
+  }
+
+  public static Map<String, String> generateAuthProperties(String serviceAccount,
+                                                           boolean isServiceAccountFilePath,
+                                                           String... keyPrefix) throws IOException {
+    Map<String, String> properties = new HashMap<>();
+    String privateKeyData = null;
+    properties.put(SERVICE_ACCOUNT_TYPE, isServiceAccountFilePath ? SERVICE_ACCOUNT_TYPE_FILE_PATH
+      : SERVICE_ACCOUNT_TYPE_JSON);
+
+    for (String prefix : keyPrefix) {
+      if (isServiceAccountFilePath) {
+        properties.put(String.format("%s.%s", prefix, CLOUD_JSON_KEYFILE_SUFFIX), serviceAccount);
+        continue;
+      }
+      ServiceAccountCredentials credentials = loadServiceAccountCredentials(serviceAccount, false);
+
+      properties.put(String.format("%s.%s", prefix, CLOUD_ACCOUNT_EMAIL_SUFFIX), credentials.getClientEmail());
+      properties.put(String.format("%s.%s", prefix, CLOUD_ACCOUNT_PRIVATE_KEY_ID_SUFFIX),
+                     credentials.getPrivateKeyId());
+      if (privateKeyData == null) {
+        privateKeyData = extractPrivateKey(credentials);
+      }
+      properties.put(String.format("%s.%s", prefix, CLOUD_ACCOUNT_KEY_SUFFIX), privateKeyData);
+      properties.put(String.format("%s.%s", prefix, CLOUD_ACCOUNT_JSON_SUFFIX), serviceAccount);
+    }
+    return properties;
+  }
+
   public static Map<String, String> getFileSystemProperties(GCPConfig config, String path,
                                                             Map<String, String> properties) {
-    String serviceAccountFilePath = config.getServiceAccountFilePath();
-    if (serviceAccountFilePath != null) {
-      properties.put(CLOUD_JSON_KEYFILE, serviceAccountFilePath);
+    try {
+      properties.putAll(generateAuthProperties(config.getServiceAccount(), config.isServiceAccountFilePath(),
+                                               CLOUD_JSON_KEYFILE_PREFIX));
```

Comment:
why do we need to pass this? There is no need of this prefix.

Reply:
This is used in following places:

https://github.com/data-integrations/google-cloud/blob/ff04f4783a3bd82dafb9d87bff150f6061bab322/src/main/java/io/cdap/plugin/gcp/bigquery/util/BigQueryUtil.java#L139


https://github.com/data-integrations/google-cloud/blob/48a86cecc1cdf7e45b95bfbe76b19662754b8f2a/src/main/java/io/cdap/plugin/gcp/bigquery/util/BigQueryUtil.java#L155-L158